File: internal/relay/nip17.go
package relay

import (
	"encoding/hex"
	"encoding/json"
	"fmt"

	kcrypto "github.com/opc/keychat/internal/crypto"
)

type Rumor struct {
	Content   string `json:"content"`
	Kind      int    `json:"kind"`
	Tags      []Tag  `json:"tags"`
	CreatedAt int64  `json:"created_at"`
	Pubkey    string `json:"pubkey"`
}

type Seal struct {
	Content   string `json:"content"`
	Kind      int    `json:"kind"`
	Tags      []Tag  `json:"tags"`
	CreatedAt int64  `json:"created_at"`
	Pubkey    string `json:"pubkey"`
}

func CreateGiftWrap(rumorContent string, senderPrivKey, recipientPubKey []byte) (*Event, error) {
	senderPubKey := kcrypto.GetPubkey(senderPrivKey)

	rumor := Rumor{
		Content:   rumorContent,
		Kind:      14,
		Tags:      []Tag{{"p", hex.EncodeToString(recipientPubKey)}},
		CreatedAt: 1700000000,
		Pubkey:    hex.EncodeToString(senderPubKey),
	}
	rumorJSON, err := json.Marshal(rumor)
	if err != nil {
		return nil, fmt.Errorf("marshal rumor: %w", err)
	}

	encryptedRumor, err := kcrypto.Encrypt(string(rumorJSON), senderPrivKey, recipientPubKey)
	if err != nil {
		return nil, fmt.Errorf("encrypt rumor: %w", err)
	}

	seal := Seal{
		Content:   encryptedRumor,
		Kind:      13,
		Tags:      []Tag{{"p", hex.EncodeToString(senderPubKey)}},
		CreatedAt: 1700000001,
		Pubkey:    hex.EncodeToString(senderPubKey),
	}
	sealJSON, err := json.Marshal(seal)
	if err != nil {
		return nil, fmt.Errorf("marshal seal: %w", err)
	}

	ephemeralPriv, _, err := kcrypto.GenerateKey()
	if err != nil {
		return nil, fmt.Errorf("generate ephemeral key: %w", err)
	}

	encryptedSeal, err := kcrypto.Encrypt(string(sealJSON), ephemeralPriv, recipientPubKey)
	if err != nil {
		return nil, fmt.Errorf("encrypt seal: %w", err)
	}

	ephemeralPub := kcrypto.GetPubkey(ephemeralPriv)
	giftWrap := &Event{
		Pubkey:    hex.EncodeToString(ephemeralPub),
		CreatedAt: 1700000002,
		Kind:      1059,
		Tags:      []Tag{{"p", hex.EncodeToString(recipientPubKey)}},
		Content:   encryptedSeal,
	}
	giftWrap.ID = giftWrap.ComputeID()

	return giftWrap, nil
}

func UnwrapGiftWrap(giftWrap *Event, recipientPrivKey []byte) (*Rumor, error) {
	if giftWrap.Kind != 1059 {
		return nil, fmt.Errorf("not a gift wrap: kind %d", giftWrap.Kind)
	}

	ephemeralPub, err := hex.DecodeString(giftWrap.Pubkey)
	if err != nil {
		return nil, fmt.Errorf("decode ephemeral pubkey: %w", err)
	}

	sealJSON, err := kcrypto.Decrypt(giftWrap.Content, recipientPrivKey, ephemeralPub)
	if err != nil {
		return nil, fmt.Errorf("decrypt gift wrap: %w", err)
	}

	var seal Seal
	if err := json.Unmarshal([]byte(sealJSON), &seal); err != nil {
		return nil, fmt.Errorf("parse seal: %w", err)
	}

	senderPubKey, err := hex.DecodeString(seal.Pubkey)
	if err != nil {
		return nil, fmt.Errorf("decode sender pubkey: %w", err)
	}

	rumorJSON, err := kcrypto.Decrypt(seal.Content, recipientPrivKey, senderPubKey)
	if err != nil {
		return nil, fmt.Errorf("decrypt seal: %w", err)
	}

	var rumor Rumor
	if err := json.Unmarshal([]byte(rumorJSON), &rumor); err != nil {
		return nil, fmt.Errorf("parse rumor: %w", err)
	}

	return &rumor, nil
}

func GetGiftWrapRecipients(event *Event) []string {
	if event.Kind != 1059 {
		return nil
	}
	var recipients []string
	for _, tag := range event.Tags {
		if len(tag) > 1 && tag[0] == "p" {
			recipients = append(recipients, tag[1])
		}
	}
	return recipients
}

File: web/src/crypto/nip17.ts
import { encrypt, decrypt } from './nip44'
import { getPublicKey } from './keys'
import { bytesToHex } from '@noble/hashes/utils'
import { secp256k1 } from '@noble/curves/secp256k1'

export interface GiftWrap {
  kind: 1059
  pubkey: string
  content: string
  tags: string[][]
}

export function createGiftWrap(
  rumorContent: string,
  senderPrivKey: Uint8Array,
  recipientPubKey: Uint8Array,
): GiftWrap {
  const senderPubKey = getPublicKey(senderPrivKey)

  const rumor = JSON.stringify({
    content: rumorContent,
    kind: 14,
    tags: [['p', bytesToHex(recipientPubKey)]],
    pubkey: bytesToHex(senderPubKey),
  })

  const encryptedRumor = encrypt(rumor, senderPrivKey, recipientPubKey)

  const seal = JSON.stringify({
    content: encryptedRumor,
    kind: 13,
    tags: [['p', bytesToHex(senderPubKey)]],
    pubkey: bytesToHex(senderPubKey),
  })

  const ephemeralPriv = secp256k1.utils.randomPrivateKey()
  const ephemeralPub = getPublicKey(ephemeralPriv)

  const encryptedSeal = encrypt(seal, ephemeralPriv, recipientPubKey)

  return {
    kind: 1059,
    pubkey: bytesToHex(ephemeralPub),
    content: encryptedSeal,
    tags: [['p', bytesToHex(recipientPubKey)]],
  }
}

export function unwrapGiftWrap(
  giftWrap: GiftWrap,
  recipientPrivKey: Uint8Array,
): string {
  if (giftWrap.kind !== 1059) {
    throw new Error(`not a gift wrap: kind ${giftWrap.kind}`)
  }

  const ephemeralPub = hexToBytes(giftWrap.pubkey)
  const sealJSON = decrypt(giftWrap.content, recipientPrivKey, ephemeralPub)
  const seal = JSON.parse(sealJSON)

  const senderPubKey = hexToBytes(seal.pubkey)
  const rumorJSON = decrypt(seal.content, recipientPrivKey, senderPubKey)
  const rumor = JSON.parse(rumorJSON)

  return rumor.content
}

function hexToBytes(hex: string): Uint8Array {
  const bytes = new Uint8Array(hex.length / 2)
  for (let i = 0; i < bytes.length; i++) {
    bytes[i] = parseInt(hex.substr(i * 2, 2), 16)
  }
  return bytes
}
